packages/arb-bridge-eth/contracts/rollup/Rollup.sol

```diff
@@ -0,0 +1,699 @@
+        // This is safe because it occurs after all checks and effects
+        sendStakerFunds(destination, amount);
+        return amount;
+    }
+
+    /**
+     * @notice Move stake onto an existing node
+     * @param blockHash Hash of a recent block to protect against reorgs
+     * @param blockNumber Block number with the given hash
+     * @param nodeNum Inbox of the node to move stake to. This must by a child of the node the staker is currently staked on
+     */
+    function stakeOnExistingNode(
+        bytes32 blockHash,
+        uint256 blockNumber,
+        uint256 nodeNum
+    ) external whenNotPaused {
+        require(isStaked(msg.sender), "NOT_STAKED");
+
+        require(blockhash(blockNumber) == blockHash, "invalid known block");
+        require(nodeNum >= firstUnresolvedNode() && nodeNum <= latestNodeCreated());
+        INode node = getNode(nodeNum);
+        require(latestStakedNode(msg.sender) == node.prev(), "NOT_STAKED_PREV");
+        stakeOnNode(msg.sender, nodeNum);
+    }
+
+    /**
+     * @notice Move stake onto a new node
+     * @param blockHash Hash of a recent block to protect against reorgs
+     * @param blockNumber Block number with the given hash
+     * @param assertionBytes32Fields Assertion data for creating
+     * @param assertionIntFields Assertion data for creating
+     */
+    function stakeOnNewNode(
+        bytes32 blockHash,
+        uint256 blockNumber,
+        uint256 nodeNum,
+        bytes32[7] calldata assertionBytes32Fields,
+        uint256[10] calldata assertionIntFields
+    ) external whenNotPaused {
+        require(isStaked(msg.sender), "NOT_STAKED");
+        require(blockhash(blockNumber) == blockHash, "invalid known block");
+        require(nodeNum == latestNodeCreated() + 1, "NODE_NUM");
+        RollupLib.Assertion memory assertion =
+            RollupLib.decodeAssertion(assertionBytes32Fields, assertionIntFields);
+        INode prevNode = getNode(latestStakedNode(msg.sender));
+        // Make sure the previous state is correct against the node being built on
+        require(
+            RollupLib.beforeNodeStateHash(assertion) == prevNode.stateHash(),
+            "PREV_STATE_HASH"
+        );
+
+        // inboxMaxCount must be greater than beforeInboxCount since we can't have read past the end of the inbox
+        (uint256 inboxMaxCount, bytes32 inboxMaxAcc) = bridge.inboxInfo();
+        require(
+            assertion.inboxMessagesRead <= inboxMaxCount - assertion.beforeInboxCount,
+            "INBOX_PAST_END"
+        );
+
+        uint256 timeSinceLastNode = block.number - assertion.beforeProposedBlock;
+        // Verify that assertion meets the minimum Delta time requirement
+        require(timeSinceLastNode >= minimumAssertionPeriod(), "TIME_DELTA");
+
+        // Minimum size requirements: each assertion must satisfy either
+        require(
+            // Consumes at least all inbox messages put into L1 inbox before your prev node’s L1 blocknum
```

Comment:
Should have a third possibility here, which is that the assertion has the maximum allowed number of sends.  (Also, should enforce a maximum allowed number of sends.)

Reply:
Added with  a todo to figure out what the maximum should be